framework/include/materials/FunctorMaterialProperty.h

```diff
@@ -0,0 +1,192 @@
+    if (!pr.second)
+      mooseError("No insertion for the functor material property '",
+                 _name,
+                 "' for block id ",
+                 block_id,
+                 ". Another material must already declare this property on that block.");
+    _elem_and_face_functor.emplace(block_id, my_lammy);
+    _face_functor.emplace(block_id, my_lammy);
+    _qp_functor.emplace(block_id, my_lammy);
+    _tqp_functor.emplace(block_id, my_lammy);
+  };
+
+  for (const auto block_id : block_ids)
+  {
+    if (block_id == Moose::ANY_BLOCK_ID)
+    {
+      const auto & inner_block_ids = mesh.meshSubdomains();
+      for (const auto inner_block_id : inner_block_ids)
+        add_lammy(inner_block_id);
+    }
+    else
+      add_lammy(block_id);
+  }
+}
+
+template <typename T>
+std::string
+FunctorMaterialProperty<T>::subdomainErrorMessage(const SubdomainID sub_id) const
+{
+  return "The provided subdomain ID " + std::to_string(sub_id) +
+         " doesn't exist in the map for material property " + _name +
+         "! This is likely because you did not provide a functor material "
+         "definition on that subdomain";
+}
+
+template <typename T>
+T
+FunctorMaterialProperty<T>::evaluate(const Elem * const & elem, unsigned int state) const
+{
+  mooseAssert(elem && elem != libMesh::remote_elem,
+              "The element must be non-null and non-remote in functor material properties");
+  auto it = _elem_functor.find(elem->subdomain_id());
+  mooseAssert(it != _elem_functor.end(), subdomainErrorMessage(elem->subdomain_id()));
+  return it->second(elem, state);
+}
+
+template <typename T>
+T
+FunctorMaterialProperty<T>::evaluate(const ElemAndFaceArg & elem_and_face, unsigned int state) const
+{
+  mooseAssert((std::get<0>(elem_and_face) && std::get<0>(elem_and_face) != libMesh::remote_elem) ||
+                  std::get<1>(elem_and_face),
+              "The element must be non-null and non-remote or the face must be non-null in functor "
+              "material properties");
+  auto it = _elem_and_face_functor.find(std::get<2>(elem_and_face));
+  mooseAssert(it != _elem_and_face_functor.end(),
+              subdomainErrorMessage(std::get<2>(elem_and_face)));
+  return it->second(elem_and_face, state);
+}
+
+template <typename T>
+T
+FunctorMaterialProperty<T>::evaluate(const FaceArg & face, unsigned int state) const
+{
+  mooseAssert(std::get<0>(face), "FaceInfo must be non-null");
+  auto it = _face_functor.find(std::get<3>(face));
+  mooseAssert(it != _face_functor.end(), subdomainErrorMessage(std::get<3>(face)));
+  return it->second(face, state);
+}
+
+template <typename T>
+T
+FunctorMaterialProperty<T>::evaluate(const QpArg & elem_and_qp, unsigned int state) const
+{
+  auto it = _qp_functor.find(std::get<0>(elem_and_qp)->subdomain_id());
+  mooseAssert(it != _qp_functor.end(),
+              subdomainErrorMessage(std::get<0>(elem_and_qp)->subdomain_id()));
+  return it->second(elem_and_qp, state);
```

Comment:
I don't think we should treat this overload any differently than the others in this particular class template. The purpose is on-the-fly material property `evaluate`ion. If we end up being forwarded onto a `Function` or a `MooseVariable` then they have implemented checks on whether we've already reinit'd certain data. A given material property, however, has no idea what the actual user lambda implementation is and so I don't think it can assume anything

Reply:
A user may choose to cache material property evaluation in which case if we've already computed for a given `elem_and_qp`, then we will not call this `evaluate` overload; instead we'll stay entirely in the non-virtual `operator()` implementation in the `FunctorInterface` base template